### PR TITLE
*: provide a configurable tidb_retry_limit session variable

### PR DESCRIPTION
**Pull request by tiancaiamao**
tidb_retry_limit variable affects transaction retry count, if it's
set to 0, transaction will not retry.

@coocood @jackysp

### CODE REVIEW

**Comment by ngaut** [comment]
Good job.

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -96,6 +96,9 @@ const (
 	// tidb_general_log is used to log every query in the server in info level.
 	TiDBGeneralLog = "tidb_general_log"
 
+	// tidb_retry_limit is the maximun number of retries when commit a transaction.
```
Comment (shenli) [inline sessionctx/variable/tidb_vars.go]: when commiting

Comment (shenli) [reply sessionctx/variable/tidb_vars.go]: Please comment for the affected scope of this variable.

**Comment by shenli** [comment]
Please remind to update the document.

File: config/config.go [hunk 2]
```diff
@@ -143,7 +143,7 @@ type Status struct {
 type Performance struct {
 	MaxProcs            uint    `toml:"max-procs" json:"max-procs"`
 	TCPKeepAlive        bool    `toml:"tcp-keep-alive" json:"tcp-keep-alive"`
-	RetryLimit          uint    `toml:"retry-limit" json:"retry-limit"`
+	RetryLimit          uint32  `toml:"retry-limit" json:"retry-limit"`
```
Comment (coocood) [inline config/config.go]: Why change the type to `uint32`?

File: config/config_test.go [hunk 3]
```diff
@@ -56,7 +56,7 @@ commit-timeout="41s"`)
 	c.Assert(conf.BinlogSocket, Equals, "/tmp/socket")
 
 	// Test that the value will be overwritten by the config file.
-	c.Assert(conf.Performance.RetryLimit, Equals, uint(10))
+	c.Assert(conf.Performance.RetryLimit, Equals, uint32(10))
```
Comment (zhexuany) [inline config/config_test.go]: I have same concern with Yuxing. Why do we change `unit` to `unit32`? The only reason I can come up with is `uint32`  need less space. Is my guess be the reason?

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**